libyul/optimiser/ConstantOptimiser.cpp

```diff
@@ -0,0 +1,171 @@
+{
+	if (_e.type() == typeid(Literal))
+	{
+		Literal const& literal = boost::get<Literal>(_e);
+		if (literal.kind != LiteralKind::Number)
+			return;
+
+		if (
+			Expression const* repr =
+				RepresentationFinder(m_evmVersion, m_meter, locationOf(_e), m_cache)
+				.tryFindRepresentation(valueOfLiteral(literal))
+		)
+			_e = ASTCopier{}.translate(*repr);
+	}
+	else
+		ASTModifier::visit(_e);
+}
+
+Expression const* RepresentationFinder::tryFindRepresentation(dev::u256 const& _value)
+{
+	if (_value < 0x10000)
+		return nullptr;
+
+	Representation const& repr = findRepresentation(_value);
+	if (repr.expression->type() == typeid(Literal))
+		return nullptr;
+	else
+		return repr.expression.get();
+}
+
+Representation const& RepresentationFinder::findRepresentation(dev::u256 const& _value)
+{
+	if (m_cache.count(_value))
+		return m_cache.at(_value);
+
+	Representation routine = represent(_value);
+
+	if (dev::bytesRequired(~_value) < dev::bytesRequired(_value))
+		// Negated is shorter to represent
+		routine = min(move(routine), represent("not", findRepresentation(~_value)));
+
+	// Decompose value into a * 2**k + b where abs(b) << 2**k
+	for (unsigned bits = 255; bits > 8 && m_maxSteps > 0; --bits)
+	{
+		unsigned gapDetector = unsigned((_value >> (bits - 8)) & 0x1ff);
+		if (gapDetector != 0xff && gapDetector != 0x100)
+			continue;
+
+		u256 powerOfTwo = u256(1) << bits;
+		u256 upperPart = _value >> bits;
+		bigint lowerPart = _value & (powerOfTwo - 1);
+		if ((powerOfTwo - lowerPart) < lowerPart)
+		{
+			lowerPart = lowerPart - powerOfTwo; // make it negative
+			upperPart++;
+		}
+		if (upperPart == 0)
+			continue;
+		if (abs(lowerPart) >= (powerOfTwo >> 8))
+			continue;
+		Representation newRoutine;
+		if (m_evmVersion.hasBitwiseShifting())
+			newRoutine = represent("shl", represent(bits), findRepresentation(upperPart));
+		else
+		{
+			newRoutine = represent("exp", represent(2), represent(bits));
+			if (upperPart != 1)
+				newRoutine = represent("mul", findRepresentation(upperPart), newRoutine);
+		}
+
+		if (newRoutine.cost >= routine.cost)
+			continue;
+
+		if (lowerPart > 0)
+			newRoutine = represent("add", newRoutine, findRepresentation(u256(abs(lowerPart))));
+		else if (lowerPart < 0)
+			newRoutine = represent("sub", newRoutine, findRepresentation(u256(abs(lowerPart))));
+
+		if (m_maxSteps > 0)
+			m_maxSteps--;
+		routine = min(move(routine), move(newRoutine));
+	}
+	return m_cache[_value] = move(routine);
+}
+
+Representation RepresentationFinder::represent(dev::u256 const& _value) const
+{
+	Representation repr;
+	repr.expression = make_unique<Expression>(Literal{m_location, LiteralKind::Number, YulString{formatNumber(_value)}, {}});
+	repr.cost = m_meter.costs(*repr.expression);
+	return repr;
+}
+
+Representation RepresentationFinder::represent(
+	string const& _instruction,
+	Representation const& _argument
+) const
+{
+	dev::eth::Instruction instr = Parser::instructions().at(_instruction);
```

Comment:
Actually, is there any reason for involving the parser here instead of just having a ``dev::eth:::Instruction`` as argument and using that directly?

Reply:
Yep, that should probably be factored out.

Reply:
Hm, we can also merge #6775  first...